Transliterate Greek according to ELOT 743. Untested

SStelioss opened the pull request:
This is as far as I'm gonna go today, bit of a noob in python and I hope it doesn't show *too* much. Doesn't work properly, and I have no idea why, but I figure I'm pretty close? And at least it doesn't throw an error! Probably... It *should* be working. Anyway. Heavily WIP, consider merging after testing.
### Current limitations
Can't do accents according to the standard, expects a de-accented input. ~~Doesn't work.~~ Possible redundancy in the ifs. Suggestions to improve welcome!

It does not sanity-check inputs or outputs, and has no testing it can run to make sure future edits don't break something.

Commit:
WIP: Doesn't work properly.

SStelioss commented:
Oh my god. Right after I did the pull request, I figured out why it doesn't work; I replace the thing just fine, but I never actually commit any of my changes to a variable...

EDIT: Nope, that's not it. The if statements after the for loops appear to not be called at all

Commit:
FIXED IT! SHOULD WORK FINE!

SStelioss commented:
Works fine. I had three mistakes I fixed:

1. Didn't properly declare my lists
2. Couldn't handle ς, used a placeholder fix (replace all instances of ς with σ before processing, very placeholder)
3. Hadn't actually comitted the string.replace to a variable

More testing is required, but I think it works fine.

Commit:
FIXED IT! SHOULD WORK FINE!

Commit:
Bugfixing, cleaned up, added mixed-casing support (Ψαρι is Psari, not PSari, but ΨΑΡΙ is still PSARI)

SStelioss commented:
Tested it a bit, seems to work fine. Possible point of failure is more than one word of input with two starting "Μπ"s (Μπαμπας Μπαμπουλας)

sinnec commented on src/greeklt/greek_iso_transliteration.py:
```diff
@@ -0,0 +1,272 @@
+def greek_iso_transliteration(string: str):
```
I would rather name this `greek_elot_transliteration`

SStelioss replied:
Fixed!

sinnec commented on src/greeklt/greek_iso_transliteration.py:
```diff
@@ -0,0 +1,272 @@
+def greek_iso_transliteration(string: str):
+#   This should all probably be a CSV or something
+    el_low = [
```
Instead of lists, it would be prefferable for readabilty reasons to store our letters in dictionaries, eg. {"β": "b"}. And they could all be moved in a seperate file later on, along with other functions' dictionaries.

SStelioss replied:
Done! Letters are in a dictionary; as a result I was able to simplify the code a lot. Thanks for the advice! I'd really love to quote some relevant pieces of code but I have no idea how; sorry!

sinnec commented on src/greeklt/greek_iso_transliteration.py:
```diff
@@ -0,0 +1,272 @@
+        "Χ",
+        "Ψ"
+    ]
+#   Accent based digraphs
+#    el_low_acc_digraphs = [
+#        "άυ",
+#        "αϋ",
+#
+#        "έυ",
+#        "εϋ",
+#
+#        "ήυ",
+#        "ηϋ"
+#
+#    ]
+#    el_mix_acc_digraphs = [
+#        "Άυ",
+#        "Αϋ",
+#
+#        "Έυ",
+#      "Εϋ",
+#
+#        "Ήυ",
+#        "Ηϋ"
+#    ]
+#    el_cap_acc_digraphs = [
+#        "ΆΥ",
+#        "ΑΫ",
+#
+#        "ΈΥ",
+#        "ΕΫ",
+#
+#        "ΉΥ",
+#        "ΗΫ"
+#    ]
```
As for accents, I think we could drop all accentuation like here: https://www.passport.gov.gr/passports/GrElotConverter/GrElotConverter.html

From `ψάρι` to `psari` and not `psάri`.

SStelioss replied:
For now, in the interests of getting it finished I decided to fully drop the accentuation as a feature. However the ELOT 743 standard actually does have a table for non-character conversions, and ultimately I think we should conform to the standard as close as possible. Maybe have accent conversion be an option or something.

sinnec replied:
The standard has the accentuation but my intention was to drop the altogether because I haven't ever found a use for it. Maybe we could implement it later!

SStelioss replied:
Fair! It's a very limited use case and even theoretically I can't see an actual use for it. It'd be a fun side project though.

Safe to say we've resolved this issue as a future TBD; I'll leave the commented out lines in place to expand upon them in the future

sinnec reviewed (commented):
Good work. I would expect a few tests before merging. I could write the docs for that!

Commit:
Bugfixing, cleaned up, added mixed-casing support (Ψαρι is Psari, not PSari, but ΨΑΡΙ is still PSARI)

Commit:
Added capability for transliterating entire sentences. Fixed bugs that arose from that. Made lists into dictionary and simplified the "dumb" transliteration code thanks to that. Used remove_accentuation to allow for input of accented sentences. Added proper transliteration for "ου" in all cases accented or unaccented. To facilitate that, changed remove_accentuation.py to have an optional input that skips the diairesis step

SStelioss committed:
Merge branch 'sinnec:main' into main

Commit:
In remove_accentuation.py: Made readability & common sense changes, added features as required for greek_elot_transliteration.py

In greek_elot_transliteration.py: Fixed bugs revolving around ου, made some mild readability improvements

In README.md: Minor grammar changes, documented changes to remove_accentuation.py as well as usage of greek_elot_transliteration.py

In test_tests.py: Added rudimentary tests. Not sure if they're properly implemented.

Commit:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	README.md

Commit:
Fixed merge conflicts in a very weird way. I need to get better with git.

SStelioss commented:
> Good work. I would expect a few tests before merging. I could write the docs for that!

I think I did the tests? I wrote them, at the very least. I've never written tests before hahah. Would appreciate some feedback.

Speaking of docs, I wrote some! My branch conflicted with yours, and I had no idea how to properly merge 'em so I improvised a little. Def not a good way of doing things, but ah well.

I'm thinking it's in a state where it could be merged, but I'd appreciate your feedback on all the changes I made, and if there's any way I can improve it.

sinnec commented:
Sorry for the delay but couldn't find time to review it. I'll test it and leave a comment at the code section and a suggestion! But great work altogether!

sinnec commented on src/greeklt/greek_elot_transliteration.py:
```diff
@@ -0,0 +1,231 @@
+        "CH",
+        "PS"
+    ]
+#   Accent based digraphs
+#    el_low_acc_digraphs = [
+#        "άυ",
+#        "αϋ",
+#
+#        "έυ",
+#        "εϋ",
+#
+#        "ήυ",
+#        "ηϋ"
+#
+#    ]
+#    el_mix_acc_digraphs = [
+#        "Άυ",
+#        "Αϋ",
+#
+#        "Έυ",
+#        "Εϋ",
+#
+#        "Ήυ",
+#        "Ηϋ"
+#    ]
+#    el_cap_acc_digraphs = [
+#        "ΆΥ",
+#        "ΑΫ",
+#
+#        "ΈΥ",
+#        "ΕΫ",
+#
+#        "ΉΥ",
+#        "ΗΫ"
+#    ]
```
I suggest we delete all commented code since it's not used!

SStelioss replied:
Hmm... Reminder to myself not to use GitHub on the phone again. It had me fooled, thinking I only had one notification haha

Sure! I'll keep it saved in a file or something on my own computer for future use.

sinnec commented on src/greeklt/remove_accentuation.py:
```diff
@@ -1,4 +1,4 @@
-def remove_accentuation(string: str):
+def remove_accentuation(string: str, modifier=0):
```
It should be modifier=True not 0. I also would like to rename modifier to something like keep_dieresis=True or something like that.

sinnec commented on src/greeklt/remove_accentuation.py:
```diff
@@ -17,14 +17,19 @@ def remove_accentuation(string: str):
             char = accents[c]
-        if c in dieresis.keys() and prev_char in ("ά", "ό", "έ"):
-            char = dieresis[c]
+        if modifier == 0:
+            if c in dieresis.keys() and prev_char in ("ά", "ό", "έ"):
+                char = dieresis[c]
+        if modifier == 1:  # Remove dieresis
+            if c in dieresis_reverse.keys():
+                char = dieresis_reverse[c]
```
if modifier == 0 can become `if modifier:`which equals to `if modifier == True`,
and if modifier == 1 can become `if not modifier:` which equals to `if modifier == False`.

sinnec commented on src/greeklt/greek_elot_transliteration.py:
```diff
@@ -0,0 +1,231 @@
+        "φ",
+        "χ",
+        "ψ"
+    ]
```
I would preffer all list pairs tyo be converted into dicts but it's more of a readability enchancement!

SStelioss commented:
> Sorry for the delay but couldn't find time to review it. I'll test it and leave a comment at the code section and a suggestion! But great work altogether!

No problem at all, and thanks! Take your time, there's no rush!

sinnec commented:
Hey! Is there any progress?

SStelioss commented:
Hey! Sorry for not being active. I've been swamped with schoolwork, but once the easter break starts in a few weeks I plan to pick the project back up. Thanks for your understanding!
